libcudacxx/include/cuda/__floating_point/operators.h

```diff
@@ -0,0 +1,1250 @@
+  {
+    return _Rhs{};
+  }
+  else
+  {
+    static_assert(__always_false<sizeof(_Lhs)>(), "Cannot make a common fp type from the given types");
+    _CCCL_UNREACHABLE();
+  }
+}
+
+template <class _Lhs, typename _Rhs>
+using __fp_common_type_t = decltype(__fp_make_common_type<_Lhs, _Rhs>());
+
+// Implementations of the arithmetic operations. Usually an operation is in several parts:
+//   1. __op() - the entry point for the operation which tries to implement the operation using the host & device native
+//      types if available, otherwise calls the __op_impl() function.
+//
+//   2. __op_impl() - the non native implementation of the operation for the given type. This function dispatches to the
+//      target specific implementation and falls back to the constexpr implementation.
+//
+//   3. __op_impl_TARGET() - the target specific implementation. If available, implements the operation via asm.
+//
+//   4. __op_impl_constexpr() - the constexpr implementation of the operation. Slow.
+//
+// The arguments may be heterogenous. In that case the implementation is chosen in the following order:
+//   1. try to use the host & device native types
+//   2. try to use host & device native instructions for mixed arithmetic
+//   3. cast the arguments to the common type and use the implementation for homogeneous arguments
+struct __fp_ops
+{
+  /********************************************************************************************************************/
+  // Negation
+  /********************************************************************************************************************/
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr static _Tp __neg_impl_constexpr(const _Tp& __src) noexcept
+  {
+    auto __ret{__src};
+    __ret.__set_sign(!__ret.__get_sign());
+    return __ret;
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE static _Tp __neg_impl_device(const _Tp& __src) noexcept
+  {
+    [[maybe_unused]] _Tp __ret;
+
+#  if __cccl_ptx_isa >= 600
+    if constexpr (_CUDA_VSTD::_CCCL_TRAIT(is_same, _Tp, fp16))
+    {
+      NV_IF_TARGET(NV_PROVIDES_SM_53,
+                   (asm("neg.f16 %0, %1;" : "=h"(__ret.__storage_) : "h"(__src.__storage_)); return __ret;))
+    }
+#  endif // __cccl_ptx_isa >= 600
+#  if __cccl_ptx_isa >= 700
+    if constexpr (_CUDA_VSTD::_CCCL_TRAIT(is_same, _Tp, bf16))
+    {
+      NV_IF_TARGET(NV_PROVIDES_SM_80,
+                   (asm("neg.bf16 %0, %1;" : "=h"(__ret.__storage_) : "h"(__src.__storage_)); return __ret;))
+    }
+#  endif // ^^^ __cccl_ptx_isa < 700 ^^^
+
+    return __neg_impl_constexpr(__src);
+  }
+
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __neg_impl(const _Tp& __src) noexcept
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      NV_IF_TARGET(NV_IS_DEVICE, (return __neg_impl_device(__src);))
+    }
```

Comment:
I had to test if this would work with NVC++.  It does.  `NV_IF_TARGET` doesn't prevent the function from being `constexpr` if it is inside a `!is_constant_evaluated()` block.